File: src/core/prepare.ts
import { validateConceptRequest } from './concept.js';
import { resolveConceptRequest, validateConceptDefinition } from './concept-definition.js';
import { withDefinitionEvidence, definitionProgression } from './blueprint/definition.js';
import { authorEvidence } from './blueprint/evidence.js';
import { getRulePack, type RulePack } from './rulepack.js';
import { assertReferencePackCoherent, danglingReferenceEvidence } from './reference.js';
import { validateLayoutPlan } from './design.js';
import { requestSchema, type AuthorRequest, type PreparedRequest } from './schemas.js';
/** Freeze cloned, validated artifacts so callers cannot change retained evidence. */
export function freeze<T>(value: T): T {
  if (value !== null && typeof value === 'object') {
    for (const child of Object.values(value)) {
      freeze(child);
    }
    Object.freeze(value);
  }
  return value;
}

export async function hashRequest(request: AuthorRequest): Promise<string> {
  const bytes = new TextEncoder().encode(JSON.stringify(requestSchema.parse(request)));
  const digest = await globalThis.crypto.subtle.digest('SHA-256', bytes);
  return `sha256:${Array.from(new Uint8Array(digest), (byte) => byte.toString(16).padStart(2, '0')).join('')}`;
}

function unique(values: readonly (string | number)[], subject: string): void {
  if (new Set(values).size !== values.length) {
    throw new Error(`${subject} must have unique identifiers or values`);
  }
}

function validateRequest(request: AuthorRequest): void {
  validateConceptDefinition(request);
  validateConceptRequest(request);
  unique(
    request.documents.map((document) => document.id),
    'Documents',
  );
  unique(
    request.constraints.map((constraint) => constraint.id),
    'Constraints',
  );
  if (!request.documents.some((document) => document.kind === 'source')) {
    throw new Error(
      'At least one supplied source document is required; a character name is insufficient evidence',
    );
  }
  if (request.previous !== null && request.feedback === null) {
    throw new Error('A revision requires explicit feedback');
  }
  validateProgression(request.progression);
  validateInterpretation(request);
  if (request.mechanicsDefinition) {
    if (
      JSON.stringify(request.progression) !==
      JSON.stringify(definitionProgression(request.mechanicsDefinition))
    )
      throw new Error(
        'Progression must match the explicit mechanics definition. Use its three paths and crosspath limits.',
      );
    if (withDefinitionEvidence(request) !== request)
      throw new Error(
        'Prepared requests must retain their mechanics definition evidence. Run prepare again.',
      );
  }
}

export async function prepareRequest(input: unknown): Promise<PreparedRequest> {
  const request = requestSchema.parse(
    withDefinitionEvidence(resolveConceptRequest(requestSchema.parse(input))),
  );
  validateRequest(request);
  return freeze({
    schemaVersion: '1',
    kind: 'prepared',
    inputHash: await hashRequest(request),
    request,
  });
}

export async function verifyPrepared(prepared: PreparedRequest): Promise<void> {
  validateRequest(prepared.request);
  if ((await hashRequest(prepared.request)) !== prepared.inputHash) {
    throw new Error('Prepared input hash does not match its retained request');
  }
}

function validateInterpretation(request: AuthorRequest): void {
  const interpretation = request.interpretation;
  if (!interpretation) return;
  if (request.mechanicsDefinition?.profile.authoringMode !== 'planned-v1')
    throw new Error('An interpretation record needs the planned-v1 authoring route.');
  assertReferencePackCoherent(interpretation.reference);
  let pack: RulePack;
  try {
    pack = getRulePack(interpretation.layout.rulePack);
  } catch {
    throw new Error(`Unknown RulePack for the interpretation: ${interpretation.layout.rulePack}.`);
  }
  if (pack.normal_progression.path_count !== 3 || pack.normal_progression.tiers_per_path !== 5)
    throw new Error(
      'The numerical backend implements 3 paths of 5 tiers. Other packs stay layout-only.',
    );
  const issues = validateLayoutPlan(interpretation.layout, interpretation.reference, pack);
  if (issues.length)
    throw new Error(
      `The interpretation record is invalid: ${issues.map((issue) => `${issue.path}: ${issue.message}`).join('; ')}`,
    );
  const known = [
    ...authorEvidence(request).map((span) => span.id),
    ...request.documents
      .filter((document) => document.kind === 'source')
      .map((document) => document.id),
  ];
  const dangling = danglingReferenceEvidence(interpretation.reference, known);
  if (dangling.length)
    throw new Error(`The interpretation cites unknown evidence: ${dangling.join(', ')}.`);
}

function validateProgression(progression: AuthorRequest['progression']): void {
  if (progression === null) {
    return;
  }
  unique(
    progression.paths.map((path) => path.id),
    'Progression paths',
  );
  for (const path of progression.paths) {
    unique(path.tiers, `Progression path ${path.id} tiers`);
  }
  if (progression.maxActivePaths > progression.paths.length) {
    throw new Error('maxActivePaths exceeds declared path count');
  }
  if (
    progression.maxPathsAboveTier !== null &&
    progression.maxPathsAboveTier.count > progression.paths.length
  ) {
    throw new Error('maxPathsAboveTier.count exceeds declared path count');
  }
  for (const combination of progression.allowedTierCombinations ?? []) {
    validateAllowedCombination(combination, progression);
  }
}

function validateAllowedCombination(
  combination: number[],
  progression: NonNullable<AuthorRequest['progression']>,
): void {
  const hasEveryPath = combination.length === progression.paths.length;
  const usesDeclaredTiers =
    hasEveryPath &&
    combination.every(
      (tier, index) => tier === 0 || progression.paths[index]!.tiers.includes(tier),
    );
  if (!usesDeclaredTiers) {
    throw new Error(
      'Allowed tier combinations must give one declared tier or zero for every path in ' +
        'declaration order',
    );
  }

  const activePathCount = combination.filter((tier) => tier > 0).length;
  const totalTiers = combination.reduce((total, tier) => total + tier, 0);
  const threshold = progression.maxPathsAboveTier;
  const exceedsActivePaths = activePathCount > progression.maxActivePaths;
  const exceedsTotalTiers =
    progression.maxTotalTiers !== null && totalTiers > progression.maxTotalTiers;
  const exceedsThreshold =
    threshold !== null &&
    combination.filter((tier) => tier > threshold.tier).length > threshold.count;
  if (exceedsActivePaths || exceedsTotalTiers || exceedsThreshold) {
    throw new Error('An allowed tier combination contradicts the supplied progression limits');
  }
}

File: src/cli/internal/unit/prepare.go
package unit

import (
	"errors"
	"fmt"
	"regexp"

	"github.com/mardwerk/unit-generator/src/cli/internal/mechanics"
	s "github.com/mardwerk/unit-generator/src/cli/internal/schema"
)

// DefinitionDocument turns a Definition into inspectable rules evidence.
func DefinitionDocument(definition mechanics.Definition) (Document, error) {
	out, issues := s.Parse(mechanics.MechanicsDefinitionSchema, s.FromGoValue(definition))
	if len(issues) > 0 {
		return Document{}, &s.Error{Issues: issues}
	}
	return Document{
		ID:   "mechanics:" + definition.ID,
		Kind: "rules",
		Text: s.Stringify(out),
		Origin: Origin{
			Location: fmt.Sprintf("unit-mechanics:%s:%s", definition.ID, definition.Revision),
			Access:   "supplied",
			Note:     strPtr("Explicit mechanics definition. Build resolution is implemented; combat simulation and balance are not certified."),
		},
	}, nil
}

// WithDefinitionEvidence appends the Definition's evidence document when missing.
// It reports whether the request changed.
func WithDefinitionEvidence(request Request) (Request, bool, error) {
	if request.MechanicsDefinition == nil {
		return request, false, nil
	}
	document, err := DefinitionDocument(*request.MechanicsDefinition)
	if err != nil {
		return request, false, err
	}
	for _, existing := range request.Documents {
		if existing.ID == document.ID {
			if s.Stringify(s.FromGoValue(existing)) != s.Stringify(s.FromGoValue(document)) {
				return request, false, errors.New("The mechanics evidence document differs from the supplied definition. Edit mechanicsDefinition, not its generated evidence document.")
			}
			return request, false, nil
		}
	}
	request.Documents = append(append([]Document(nil), request.Documents...), document)
	return request, true, nil
}

// DefinitionProgression is the request progression a Definition implies.
func DefinitionProgression(definition mechanics.Definition) Progression {
	paths := make([]ProgressionPath, 3)
	for i := range paths {
		paths[i] = ProgressionPath{ID: fmt.Sprintf("path-%d", i+1), Tiers: []int{1, 2, 3, 4, 5}}
	}
	return Progression{
		Paths:          paths,
		MaxActivePaths: definition.Progression.MaxPurchasedPaths,
		MaxPathsAboveTier: &Threshold{
			Tier:  definition.Progression.CrosspathTier,
			Count: definition.Progression.MaxAdvancedPaths,
		},
	}
}

// ParseRequest validates a request value and decodes it.
func ParseRequest(value any) (Request, error) {
	var request Request
	return request, s.ParseInto(RequestSchema, value, &request)
}

func unique[T comparable](values []T, subject string) error {
	seen := map[T]bool{}
	for _, v := range values {
		if seen[v] {
			return fmt.Errorf("%s must have unique identifiers or values", subject)
		}
		seen[v] = true
	}
	return nil
}

func validateRequest(request Request) error {
	var ids []string
	for _, d := range request.Documents {
		ids = append(ids, d.ID)
	}
	if err := unique(ids, "Documents"); err != nil {
		return err
	}
	ids = nil
	for _, c := range request.Constraints {
		ids = append(ids, c.ID)
	}
	if err := unique(ids, "Constraints"); err != nil {
		return err
	}
	hasSource := false
	for _, d := range request.Documents {
		if d.Kind == "source" {
			hasSource = true
		}
	}
	if !hasSource {
		return errors.New("At least one supplied source document is required; a character name is insufficient evidence")
	}
	if request.Previous != nil && request.Feedback == nil {
		return errors.New("A revision requires explicit feedback")
	}
	if err := validateProgression(request.Progression); err != nil {
		return err
	}
	if request.MechanicsDefinition != nil {
		expected := DefinitionProgression(*request.MechanicsDefinition)
		if request.Progression == nil || s.Stringify(s.FromGoValue(request.Progression)) != s.Stringify(s.FromGoValue(expected)) {
			return errors.New("Progression must match the explicit mechanics definition. Use its three paths and crosspath limits.")
		}
		_, changed, err := WithDefinitionEvidence(request)
		if err != nil {
			return err
		}
		if changed {
			return errors.New("Prepared requests must retain their mechanics definition evidence. Run prepare again.")
		}
	}
	return nil
}

func validateProgression(progression *Progression) error {
	if progression == nil {
		return nil
	}
	var ids []string
	for _, p := range progression.Paths {
		ids = append(ids, p.ID)
	}
	if err := unique(ids, "Progression paths"); err != nil {
		return err
	}
	for _, p := range progression.Paths {
		if err := unique(p.Tiers, "Progression path "+p.ID+" tiers"); err != nil {
			return err
		}
	}
	if progression.MaxActivePaths > len(progression.Paths) {
		return errors.New("maxActivePaths exceeds declared path count")
	}
	if progression.MaxPathsAboveTier != nil && progression.MaxPathsAboveTier.Count > len(progression.Paths) {
		return errors.New("maxPathsAboveTier.count exceeds declared path count")
	}
	if progression.AllowedTierCombinations != nil {
		for _, combination := range *progression.AllowedTierCombinations {
			if err := validateAllowedCombination(combination, progression); err != nil {
				return err
			}
		}
	}
	return nil
}

func validateAllowedCombination(combination []int, progression *Progression) error {
	usesDeclared := len(combination) == len(progression.Paths)
	if usesDeclared {
		for i, tier := range combination {
			if tier == 0 {
				continue
			}
			found := false
			for _, t := range progression.Paths[i].Tiers {
				if t == tier {
					found = true
				}
			}
			if !found {
				usesDeclared = false
			}
		}
	}
	if !usesDeclared {
		return errors.New("Allowed tier combinations must give one declared tier or zero for every path in declaration order")
	}
	active, total, above := 0, 0, 0
	for _, tier := range combination {
		if tier > 0 {
			active++
		}
		total += tier
		if progression.MaxPathsAboveTier != nil && tier > progression.MaxPathsAboveTier.Tier {
			above++
		}
	}
	if active > progression.MaxActivePaths ||
		(progression.MaxTotalTiers != nil && total > *progression.MaxTotalTiers) ||
		(progression.MaxPathsAboveTier != nil && above > progression.MaxPathsAboveTier.Count) {
		return errors.New("An allowed tier combination contradicts the supplied progression limits")
	}
	return nil
}

// Prepare validates a request value and returns a prepared request with its hash.
func Prepare(value any) (Prepared, error) {
	request, err := ParseRequest(value)
	if err != nil {
		return Prepared{}, err
	}
	request, _, err = WithDefinitionEvidence(request)
	if err != nil {
		return Prepared{}, err
	}
	if request, err = ParseRequest(s.FromGoValue(request)); err != nil {
		return Prepared{}, err
	}
	if err := validateRequest(request); err != nil {
		return Prepared{}, err
	}
	hash, err := HashRequest(request)
	if err != nil {
		return Prepared{}, err
	}
	return Prepared{SchemaVersion: "1", Kind: "prepared", InputHash: hash, Request: request}, nil
}

// VerifyPrepared checks a prepared request's rules and its hash.
func VerifyPrepared(prepared Prepared) error {
	if err := validateRequest(prepared.Request); err != nil {
		return err
	}
	ok, err := VerifyHash(prepared.Request, prepared.InputHash)
	if err != nil {
		return err
	}
	if !ok {
		return errors.New("Prepared input hash does not match its retained request")
	}
	return nil
}

var profileDocument = regexp.MustCompile(`^(default-td-profile-v[0-9]+|profile:.+|mechanics:.+)$`)

// IsProfileDocument reports whether a document ID belongs to a Profile: its
// rules or its Definition's evidence. Applying another Profile replaces them.
func IsProfileDocument(id string) bool { return profileDocument.MatchString(id) }

// ApplyProfile replaces the rules a request is generated under with a Profile:
// its task, progression, Definition and rules document. Character, sources,
// decisions and revision context stay.
func ApplyProfile(request Request, profile Profile) Request {
	progression := DefinitionProgression(profile.MechanicsDefinition)
	definition := profile.MechanicsDefinition
	request.Task = profile.Task
	request.Progression = &progression
	request.MechanicsDefinition = &definition
	var documents []Document
	for _, d := range request.Documents {
		if !IsProfileDocument(d.ID) {
			documents = append(documents, d)
		}
	}
	request.Documents = append(documents, profile.Rules)
	return request
}

// ParseProfile validates a Profile value.
func ParseProfile(value any) (Profile, error) {
	var profile Profile
	return profile, s.ParseInto(UnitProfileSchema, value, &profile)
}

// ValidateProfile checks that a request prepared under the Profile passes
// the same checks as any other.
func ValidateProfile(value any) (Profile, error) {
	profile, err := ParseProfile(value)
	if err != nil {
		return Profile{}, err
	}
	placeholder := Request{
		SchemaVersion: "1",
		Task:          profile.Task,
		Character:     Character{Name: "Profile check", Work: "Profile check", Scope: "Profile check"},
		Documents: []Document{{
			ID: "profile-check-source", Kind: "source",
			Text:   "Placeholder source text used only to validate a Profile.",
			Origin: Origin{Location: "profile-check", Access: "supplied"},
		}},
		Constraints: []Constraint{},
	}
	if _, err := Prepare(s.FromGoValue(ApplyProfile(placeholder, profile))); err != nil {
		return Profile{}, err
	}
	return profile, nil
}
